sync/statesync/code_fetcher.go

```diff
@@ -0,0 +1,239 @@
+	addWG sync.WaitGroup
+
+	// Channel of code hashes to be consumed by the code syncer.
+	codeHashes chan common.Hash
+
+	// Set of code hashes enqueued in-memory to avoid duplicate sends.
+	outstanding set.Set[common.Hash]
+
+	// Any persisted code markers found on startup to re-enqueue when initialized.
+	dbCodeHashes []common.Hash
+}
+
+type codeFetcherOptions struct {
+	autoInit       bool
+	maxOutstanding int
+}
+
+type CodeFetcherOption func(*codeFetcherOptions)
+
+// WithAutoInit toggles whether Init is called in the constructor.
+func WithAutoInit(autoInit bool) CodeFetcherOption {
+	return func(o *codeFetcherOptions) {
+		o.autoInit = autoInit
+	}
+}
+
+// WithMaxOutstandingCodeHashes overrides the queue capacity.
+func WithMaxOutstandingCodeHashes(n int) CodeFetcherOption {
+	return func(o *codeFetcherOptions) {
+		if n > 0 {
+			o.maxOutstanding = n
+		}
+	}
+}
+
+// NewCodeFetcherQueue creates a new code fetcher queue applying optional functional options.
+func NewCodeFetcherQueue(db ethdb.Database, done <-chan struct{}, opts ...CodeFetcherOption) (*CodeFetcherQueue, error) {
+	// Apply defaults then options
+	o := codeFetcherOptions{
+		autoInit:       defaultAutoInit,
+		maxOutstanding: defaultMaxOutstandingCodeHashes,
+	}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	dbCodeHashes, err := getCodeToFetchFromDB(db)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add code hashes to queue: %w", err)
+	}
+
+	q := &CodeFetcherQueue{
+		db:           db,
+		codeHashes:   make(chan common.Hash, o.maxOutstanding),
+		open:         make(chan struct{}),
+		outstanding:  set.NewSet[common.Hash](0),
+		dbCodeHashes: dbCodeHashes,
+		done:         done,
+	}
+
+	if o.autoInit {
+		if err := q.Init(); err != nil {
+			return nil, err
+		}
+	}
+
+	return q, nil
+}
+
+// CodeHashes returns the receive-only channel of code hashes to consume.
+func (q *CodeFetcherQueue) CodeHashes() <-chan common.Hash {
+	return q.codeHashes
+}
+
+// Init enqueues any persisted code markers found on disk and then marks the fetcher ready.
+func (q *CodeFetcherQueue) Init() error {
+	if err := q.addCode(q.dbCodeHashes); err != nil {
+		return fmt.Errorf("unable to resume previous sync: %w", err)
+	}
+	q.dbCodeHashes = nil
+	close(q.open)
+
+	return nil
+}
+
+// AddCode implements [synccommon.CodeFetcher] by persisting and enqueueing new hashes.
+// Blocks until the queue is initialized via Init.
+func (q *CodeFetcherQueue) AddCode(codeHashes []common.Hash) error {
+	<-q.open
+	q.addWG.Add(1)
+	defer q.addWG.Done()
+	if q.finalized.Load() {
+		return errFailedToAddCodeHashesToQueue
+	}
+
+	if len(codeHashes) == 0 {
+		return nil
+	}
+
+	return q.addCode(codeHashes)
+}
+
+// Finalize implements [synccommon.CodeFetcher] by signaling no further code hashes will be added.
+func (q *CodeFetcherQueue) Finalize() {
+	<-q.open
+	q.finalized.Store(true)
+	// Wait for any in-flight AddCode calls to complete before closing the channel.
+	q.addWG.Wait()
+	close(q.codeHashes)
```

Comment:
I'm not convinced that this protects against a race condition that will result in a send-after-close on this channel. At first I thought that I should walk through different scenarios, to prove to myself one way or the other, but that doesn't really solve the problem of the code being "obviously correct".

The golden rule for channel closing is that it must be done by either the sole sender, or the coordinator of the group of senders. Here we're in the latter scenario, but it's the coordination that I'm unsure about. I'll DM you to discuss.

Reply:
I'll give it a shot at explaining my understanding of the flow:

- `Finalize` is the coordinator - it sets `finalized=true`, then waits for all in-flight senders via `addWG.Wait()`, then closes `codeHashes`.

- Each `AddCode` "joins" the sender group with `addWG.Add(1)` before it can enqueue, and only "leaves" with the deferred `addWG.Done()` after its enqueues finish. That makes all sends happen-before the `Wait` completes.

- New `AddCode` calls after `finalized=true` are rejected (`errAddCodeAfterFinalize`) and still counted in `addWG` to avoid the narrow race where `Wait` begins just as `AddCode` arrives.

- Channel close happens only after all possible senders are either rejected or finished sending meaning there is no send-after-close path.